Strided transpose_copy.

pearu opened the pull request:
This PR introduces `transpose_copy` function for strided tensors and provides the necessary infrastructure for `transpose_copy` to support CSR and COO tensors with the aim to resolve https://github.com/pytorch/pytorch/issues/61059.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #68939
* #68938
* #68883
* #68919
* #68882
* __->__ #68867



cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

Differential Revision: [D32652525](https://our.internmc.facebook.com/intern/diff/D32652525)

pearu committed:
Strided transpose_copy.

[ghstack-poisoned]

IvanYashchuk commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -25,8 +25,24 @@
 #include <algorithm>
 #include <cstdint>
 #include <vector>
+#include <iostream>
```
This include is not needed, right?

pearu replied:
Right. Removed.

IvanYashchuk commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -25,8 +25,24 @@
+  c10::TensorOptions options = self.options();
+  DimVector sizes(self.sizes().begin(), self.sizes().end());
+  std::swap(sizes[dim0_], sizes[dim1_]);
+  set_output(sizes, options);
```
Currently, strides of the resulting tensors would be different when using `transpose` and `transpose_copy`. I think the strides should match for strided tensor inputs.

pearu replied:
Makes sense. That is, `x.transpose_copy(dim0, dim1)` is defined as `x.transpose(dim0, dim1).clone()` where `clone` preserves strides. Will fix it..

pearu replied:
Fixed.

IvanYashchuk commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -12265,6 +12272,15 @@ def ref_pairwise_distance(input1, input2):
            sample_inputs_func=sample_inputs_transpose_swapdims),
+    OpInfo('transpose_copy',
+           ref=lambda x, dim0, dim1: ref_transpose(x, dim0, dim1).copy(),
+           # transpose_copy does not support automatic differentiation for outputs with complex dtype
```
Please add "transpose_copy" to `GRADIENT_IMPLEMENTED_FOR_COMPLEX` in `pytorch/tools/autograd/gen_variable_type.py`.

pearu replied:
Done.

IvanYashchuk commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -12265,6 +12272,15 @@ def ref_pairwise_distance(input1, input2):
+           ref=lambda x, dim0, dim1: ref_transpose(x, dim0, dim1).copy(),
+           # transpose_copy does not support automatic differentiation for outputs with complex dtype
+           dtypes=all_types_and(torch.bool, torch.bfloat16, torch.half),
+           dtypesIfCUDA=all_types_and(torch.bool, torch.bfloat16, torch.half),
```
Probably `dtypesIfCUDA` is bot needed if it's the same as `dtypes`?

pearu replied:
Removed dtypesIfCUDA as unnecessary.

IvanYashchuk commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -12265,6 +12272,15 @@ def ref_pairwise_distance(input1, input2):
+           # transpose_copy does not support automatic differentiation for outputs with complex dtype
+           dtypes=all_types_and(torch.bool, torch.bfloat16, torch.half),
+           dtypesIfCUDA=all_types_and(torch.bool, torch.bfloat16, torch.half),
+           supports_out=False,
```
It should support out. Are the tests failing?

pearu replied:
`transpose_copy` does not have `out` argument defined.

IvanYashchuk replied:
`native_functions.yaml` have this defined:
```
- func: transpose_copy.out(Tensor self, int dim0, int dim1, *, Tensor(a!) out) -> Tensor(a!)
```

pearu replied:
Yes, but `out=` is still not functional:
```python
>>> import torch
>>> x = torch.tensor([[1, 2], [3, 4]])
>>> r = x.transpose_copy(0, 1)
>>> x.transpose_copy(0, 1, out=r)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: transpose_copy() got an unexpected keyword argument 'out'
```
not sure why..

pearu replied:
Ah, got it: the `out=` is available only for torch function, not for tensor methods.

pearu replied:
Removed `supports_out=False`.

pearu committed:
Update on "Strided transpose_copy."

This PR introduces `transpose_copy` function for strided tensors and provides the necessary infrastructure for `transpose_copy` to support CSR and COO tensors with the aim to resolve https://github.com/pytorch/pytorch/issues/61059.




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

pearu committed:
Update on "Strided transpose_copy."

This PR introduces `transpose_copy` function for strided tensors and provides the necessary infrastructure for `transpose_copy` to support CSR and COO tensors with the aim to resolve https://github.com/pytorch/pytorch/issues/61059.




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

cpuhrsch reviewed (approved):
LGTM

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32652525).

IvanYashchuk commented on torch/_torch_docs.py:
```diff
@@ -9920,6 +9920,35 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.transpose_copy,
+           r"""
+transpose_copy(input, dim0, dim1) -> Tensor
```
It's missing the `out` keyword argument.

pearu replied:
I presume that the doc-strings have no influence on https://github.com/pytorch/pytorch/pull/68867#discussion_r756385580 , so holding on adding `out=` to docs.

pearu replied:
Added `out=`.

IvanYashchuk commented on tools/autograd/derivatives.yaml:
```diff
@@ -1456,6 +1456,10 @@
   result: auto_linear
 
+- name: transpose_copy(Tensor self, int dim0, int dim1) -> Tensor
+  self: grad.clone().transpose_copy(dim0, dim1)
```
Why `clone()` is needed here if `transpose_copy` does the copy?

pearu replied:
Without `clone()`, few tests (test_variant_consistency_eage, test_variant_consistency_jit_transpose, test_conj_view, test_neg_view) fail with a message:
```
RuntimeError: unsupported operation: more than one element of the written-to tensor refers to a single memory location. Please clone() the tensor before performing the operation
```

IvanYashchuk replied:
Does the error appear if `at::transpose_copy` is used instead of the method?
That way we're doing two copies: one here and another inside the `transpose_copy` call.

pearu replied:
Yes, using `at::transpose_copy(grad, dim0, dim1)` leads to the same error.
However, when using `grad.clone().transpose(dim0, dim1)`, tests pass.

IvanYashchuk replied:
I think something is wrong with the `transpose_copy` then. It would be good to create a code snippet to reproduce the error.
The derivative entry now should be `self: grad.clone().transpose(dim0, dim1)` to avoid multiple copies.

pearu replied:
Here's a reproducer (when using `self: grad.transpose_copy(dim0, dim1)`):
```python
>>> x = torch.tensor([[1, 2], [3, 4]], dtype=torch.float32, requires_grad=True)
>>> y = torch.transpose_copy(x, 0, 1)
>>> y.sum().backward(retain_graph=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/pearu/git/pytorch/pytorch-transpose/torch/_tensor.py", line 352, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/pearu/git/pytorch/pytorch-transpose/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: unsupported operation: more than one element of the written-to tensor refers to a single memory location. Please clone() the tensor before performing the operation.
```

pearu replied:
The runtime error is raised when calling `out.copy_`.
Now using `self: grad.clone().transpose(dim0, dim1)`.

IvanYashchuk replied:
Is it possible to trigger the same error in normal code, without backward? Can someone potentially hit it? If yes what would be the fix?

pearu replied:
Do you mean by "normal code" when just using `transpose_copy` calls? If so, I don't have a definite answer other than test_ops.py covers testing `transpose_copy` in a number of ways.

Digging a little further, the runtime error may be a result of some transpose-state test that gives a false-positive result, after all, the `out` and `self`  in `transcopy_copy` ought to be tensors with non-overlapping storage.

Reading the message  "more than one element of the written-to tensor..", could this be related to autograd machinery in some way? @albanD , do you have any ideas?

albanD replied:
Not sure where it comes from, but why isn't the formula just: `grad.transpose(dim0, dim1)` ?

albanD replied:
Not sure where it comes from, but why isn't the formula just: `grad.transpose(dim0, dim1)` ?

pearu replied:
@albanD , see https://github.com/pytorch/pytorch/pull/68867#discussion_r756395212

albanD replied:
Ho I though this comment was when using the copy version of transpose?
What I meant here is that it is fine for the backward to be a view even if the forward is not one.

Also if you set `TORCH_SHOW_CPP_STACKTRACES=1` while running the test, do you get more details on where the error comes from?

pearu committed:
Update on "Strided transpose_copy."

This PR introduces `transpose_copy` function for strided tensors and provides the necessary infrastructure for `transpose_copy` to support CSR and COO tensors with the aim to resolve https://github.com/pytorch/pytorch/issues/61059.




cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32652525](https://our.internmc.facebook.com/intern/diff/D32652525)

[ghstack-poisoned]

pearu committed:
Update on "Strided transpose_copy."

This PR introduces `transpose_copy` function for strided tensors and provides the necessary infrastructure for `transpose_copy` to support CSR and COO tensors with the aim to resolve https://github.com/pytorch/pytorch/issues/61059.




cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32652525](https://our.internmc.facebook.com/intern/diff/D32652525)

[ghstack-poisoned]

pearu committed:
Update on "Strided transpose_copy."

This PR introduces `transpose_copy` function for strided tensors and provides the necessary infrastructure for `transpose_copy` to support CSR and COO tensors with the aim to resolve https://github.com/pytorch/pytorch/issues/61059.




cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D32652525](https://our.internmc.facebook.com/intern/diff/D32652525)

[ghstack-poisoned]

pearu commented:
Converting to draft as CSR copy transpose is implemented in `transpose`, see https://github.com/pytorch/pytorch/pull/70582
